### PR TITLE
Issue/1670 datapoints download fix part3

### PR DESCRIPTION
**Pull request by valllllll2000**
#### Before the PR (what is the issue or what needed to be done)
Na
#### The solution
incorporate the cursor param: 
* add database table + migration
* use cursor to make requests
* minor fixes after kotlin conversion
* fixed existing tests for getting datapoints
#### Screenshots (if appropriate)
NA
#### Reviewer Checklist
* [ ] Added an explanation about the work done
* [ ] Connected the PR and the issue on Zenhub
* [ ] Added a test plan to the issue
* [ ] Updated the copyright header
* [ ] Formatted the code per our style guide
* [ ] Added a documentation (if relevant)

### CODE REVIEW

**Comment by valllllll2000** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- database/src/main/java/org/akvo/flow/database/upgrade/UpgraderFactory.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Akvo/akvo-flow-mobile/pullRequest?prid=5826351&bid=19074976)

**Comment by valllllll2000** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- database/src/main/java/org/akvo/flow/database/upgrade/UpgraderFactory.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Akvo/akvo-flow-mobile/pullRequest?prid=5826351&bid=19074976)

**Comment by valllllll2000** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- database/src/main/java/org/akvo/flow/database/upgrade/UpgraderFactory.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Akvo/akvo-flow-mobile/pullRequest?prid=5826351&bid=19074976)

File: data/src/test/java/org/akvo/flow/data/repository/DataPointDataRepositoryTest.kt [hunk 1]
```diff
@@ -19,150 +19,126 @@
...
-
-        val repository =
-            DataPointDataRepository(
-                mockDataSourceFactory,
-                spyRestApi,
-                mockS3RestApi,
-                mapper,
-                mediaHelper
+    @Test
+    fun downloadDataPointsShouldReturnExpectedErrorWhenAssignmentMissing() = runBlocking {
```
Comment (muloem) [inline data/src/test/java/org/akvo/flow/data/repository/DataPointDataRepositoryTest.kt:87]: when assignment missing the returned code is a `403`? shouldnt it be `404`?

File: data/src/test/java/org/akvo/flow/data/repository/DataPointDataRepositoryTest.kt [hunk 2]
```diff
@@ -19,150 +19,126 @@
...
+        } catch (e: Exception) {
+            assert(e is AssignmentRequiredException)
+        }
     }
 
-    @Test(expected = Exception::class)
-    fun downloadDataPointsShouldReturnAnyErrorWhenNotAssignmentMissing()  = runBlockingTest {
-        doThrow(Exception()).`when`(spyRestApi).downloadDataPoints(anyLong())
+    @Test
+    fun downloadDataPointsShouldReturnExpectedErrorWhenNotAssignmentMissing() = runBlocking {
```
Comment (muloem) [inline data/src/test/java/org/akvo/flow/data/repository/DataPointDataRepositoryTest.kt:104]: its not very clear what this one means.. does that mean the assignment actually exists? if so why the exception?

File: data/src/test/java/org/akvo/flow/data/repository/DataPointDataRepositoryTest.kt [hunk 3]
```diff
@@ -19,150 +19,126 @@
...
 
-        val repository =
-            DataPointDataRepository(
-                mockDataSourceFactory,
-                spyRestApi,
-                mockS3RestApi,
-                mapper,
-                mediaHelper
+    @Test
+    fun downloadDataPointsShouldReturnAnyErrorWhenNotAssignmentMissing() = runBlocking {
```
Comment (muloem) [inline data/src/test/java/org/akvo/flow/data/repository/DataPointDataRepositoryTest.kt:121]: I also couldn't understand what kind of error this one is... i.e whats the reason that provokes it

File: data/src/main/java/org/akvo/flow/data/repository/DataPointDataRepository.kt [hunk 4]
```diff
@@ -38,11 +38,30 @@ class DataPointDataRepository @Inject constructor(
...
         try {
-            return syncDataPoints(restApi.downloadDataPoints(surveyGroupId))
+            while (moreToLoad) {
+                val apiLocaleResult = restApi.downloadDataPoints(surveyId, cursor)
+                syncedDataPoints += syncDataPoints(apiLocaleResult)
+                if (apiLocaleResult.dataPoints.isNotEmpty()) {
+                    cursor = apiLocaleResult.cursor
+                }
+                moreToLoad = apiLocaleResult.dataPoints.isNotEmpty()
+                        && apiLocaleResult.dataPoints.size >= MAX_DATAPOINT_BATCH
```
[the remark this answers is not in the document]
Comment (valllllll2000) [reply data/src/main/java/org/akvo/flow/data/repository/DataPointDataRepository.kt:59]: that's a very good point, will change it

File: data/src/main/java/org/akvo/flow/data/repository/DataPointDataRepository.kt [hunk 5]
```diff
@@ -38,11 +38,30 @@ class DataPointDataRepository @Inject constructor(
...
+    ) : DataPointRepository {
 
-    override suspend fun downloadDataPoints(surveyGroupId: Long): Int {
+    companion object {
+        private const val MAX_DATAPOINT_BATCH = 30
+    }
+
+    override suspend fun downloadDataPoints(surveyId: Long): Int {
+        var syncedDataPoints = 0
+        var cursor = dataSourceFactory.dataBaseDataSource.getDataPointCursor(surveyId)
```
[the remark this answers is not in the document]
Comment (valllllll2000) [reply data/src/main/java/org/akvo/flow/data/repository/DataPointDataRepository.kt:49]: true

File: data/src/test/java/org/akvo/flow/data/repository/DataPointDataRepositoryTest.kt (hunk 1 above)
Comment (valllllll2000) [reply data/src/test/java/org/akvo/flow/data/repository/DataPointDataRepositoryTest.kt:87]: oh right it's 404, I though it was 403 for some reason :thinking:

File: data/src/test/java/org/akvo/flow/data/repository/DataPointDataRepositoryTest.kt (hunk 3 above)
Comment (valllllll2000) [reply data/src/test/java/org/akvo/flow/data/repository/DataPointDataRepositoryTest.kt:121]: you are right this one is useless

File: data/src/test/java/org/akvo/flow/data/repository/DataPointDataRepositoryTest.kt (hunk 2 above)
Comment (valllllll2000) [reply data/src/test/java/org/akvo/flow/data/repository/DataPointDataRepositoryTest.kt:104]: any other error we just rethrow, I will rename to clarify the method

**Comment by valllllll2000** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- database/src/main/java/org/akvo/flow/database/upgrade/UpgraderFactory.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Akvo/akvo-flow-mobile/pullRequest?prid=5826351&bid=19074976)

File: data/src/test/java/org/akvo/flow/data/repository/DataPointDataRepositoryTest.kt (hunk 1 above)
Comment (muloem) [reply data/src/test/java/org/akvo/flow/data/repository/DataPointDataRepositoryTest.kt:87]: let me cross check that in the backend as well 👍

**Comment by valllllll2000** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- database/src/main/java/org/akvo/flow/database/upgrade/UpgraderFactory.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Akvo/akvo-flow-mobile/pullRequest?prid=5826351&bid=19074976)

File: data/src/test/java/org/akvo/flow/data/repository/DataPointDataRepositoryTest.kt (hunk 1 above)
Comment (muloem) [reply data/src/test/java/org/akvo/flow/data/repository/DataPointDataRepositoryTest.kt:87]: a `403` is returned when no `surveyId` is provided.  It should also probably be something like bad request `401` instead.  Is that a very big change for you?

**Pull request merged**